CPU Manager logging improvements

swatisehgal opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

We are trying to ensure consistency across resource managers when it comes to logging. While working on logging improvements for memory manager, it was identified that some parts of code base is still using klog.InfoS(..., err) instead of klog.ErrorS(err,...). This change is addressing this in CPU Manager.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Partially Fixes #123037

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

swatisehgal committed:
node: cpu-mgr: Update klog.Infof(..., err) to klog.ErrorS(err,...)

We are trying to ensure consistency across resource managers when
it comes to logging. While working on logging improvements for
memory manager, it was identified that some parts of code base
is still using klog.InfoS(..., err) instead of klog.ErrorS(err,...).
This change is addressing this in CPU Manager.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/cc @ffromani

swatisehgal committed:
node: cpu-mgr: Add logs on the happy path

While debugging, it can be useful to have logs to indicate that
things have gone as expected especially when it comes to
important events like successful startup of CPU manager
and successful allocation of resources.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: cpu-mgr: Add logs when CPU allocation is skipped

CPU Allocation is skipped in CPU Manager with static policy
in case the pod doesn't belong to Guaranteed QoS or the CPUs
requested are not integral.

We add logs to capture these skips.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -239,6 +239,8 @@ func (m *manager) Start(activePods ActivePodsFunc, sourcesReady config.SourcesRe
 
+	klog.V(4).InfoS("CPU manager started successfully", "policy", m.policy.Name())
+
 	m.allocatableCPUs = m.policy.GetAllocatableCPUs(m.state)
```
can we move the log line here?
```
klog.V(4).InfoS("CPU manager started", "policy", m.policy.Name(), "allocatableCPUs", m.allocatableCPUs)
```

I'm intentionally dropping "succesfully" because it can safely inferred implicitely ("started" + it's Info [vs Error]) and we save some arguably redundant bytes in the log entry.

swatisehgal replied:
Sure, will update.

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -264,6 +266,7 @@ func (m *manager) Allocate(p *v1.Pod, c *v1.Container) error {
 		return err
 	}
 
+	klog.V(4).InfoS("CPUs allocated by cpumanager successfully", "pod", klog.KObj(p), "podUID", p.UID, "containerName", c.Name)
```
same reasoning to drop "succesfully"
I'd rephrase as `Allocated exclusive CPUs`.
Even better: move the log line inside the `policy` so we can also log what was allocated

ffromani commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -456,6 +456,7 @@ func (p *staticPolicy) allocateCPUs(s state.State, numCPUs int, numaAffinity bit
 
 func (p *staticPolicy) guaranteedCPUs(pod *v1.Pod, container *v1.Container) int {
 	if v1qos.GetPodQOS(pod) != v1.PodQOSGuaranteed {
+		klog.V(5).InfoS("CPU Allocation Skipped for pod as it is allocated only for Guaranteed pods", "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "qos", v1qos.GetPodQOS(pod))
```
I'd compute `qos` once and use it everywhere in the function. Not for performance (I expect negligible gains) but for consistency. This way we use a single coherent value for all the function and we avoid any possible occurrence of TOCTOU issues, even though it's practically impossible because of the code works, structuring the code like that prevents that from the beginning.

I'd rephrase the message like `Exclusive CPU allocation skipped, pod QoS is not guaranteed`

swatisehgal replied:
Sure, will update.

ffromani commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -469,6 +470,7 @@ func (p *staticPolicy) guaranteedCPUs(pod *v1.Pod, container *v1.Container) int
 		}
 	}
 	if cpuQuantity.Value()*1000 != cpuQuantity.MilliValue() {
+		klog.V(5).InfoS("CPU Allocation Skipped for pod as it is allocated only for integral CPU Requests", "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "cpu", cpuQuantity.Value())
```
kinda same. Compute `cpuQuantity` once, made the log message a bit terser

ffromani reviewed (commented):
thanks for the changes! We can use some improvements in this area, so any help is welcome! Some comments inline. Nothing major, but a few points we can discuss

ffromani commented:
/triage accepted
/priority backlog

swatisehgal reviewed (commented):
Thanks, made updated as per your comments.

googs1025 commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -455,7 +456,9 @@ func (p *staticPolicy) allocateCPUs(s state.State, numCPUs int, numaAffinity bit
 }
 
 func (p *staticPolicy) guaranteedCPUs(pod *v1.Pod, container *v1.Container) int {
+	qos := v1qos.GetPodQOS(pod)
```
can we change to: 
avoid repeated calls `GetPodQOS`
```go
qos := v1qos.GetPodQOS(pod)
if qos != v1.PodQOSGuaranteed {
     // .....
}
```

swatisehgal replied:
Ah, thanks for catching that! Missed it in the previous iteration. Fixing now.

bart0sh commented on pkg/kubelet/cm/cpumanager/state/state_checkpoint.go:
```diff
@@ -201,7 +201,7 @@ func (sc *stateCheckpoint) SetCPUSet(podUID string, containerName string, cset c
 	err := sc.storeState()
 	if err != nil {
-		klog.InfoS("Store state to checkpoint error", "err", err)
+		klog.ErrorS(err, "Store state to checkpoint error", "podUID", podUID, "containerName", containerName)
```
Not directly related to this PR, but it looks a bit strange to me that `storeState` failures are not propagated up to the calling method. Any ideas why is that?

swatisehgal committed:
node: cpu-mgr: Adhere to the message style guidelines

Ensure that the log messages adhere to the message style guildelines
as captured [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#message-style-guidelines).

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

Tal-or commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -388,6 +388,7 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 	p.updateCPUsToReuse(pod, container, cpuset)
 	p.updateMetricsOnAllocate(cpuset)
 
+	klog.V(4).InfoS("Allocated exclusive CPUs", "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "cpuset", "cpuset")
```
It might be fine to have some log duplication but I'm just stating the fact the we're printing similar log at:
https://github.com/kubernetes/kubernetes/blob/cede96336a809a67546ca08df0748e4253ec270d/pkg/kubelet/cm/cpumanager/state/state_mem.go#L80

Tal-or replied:
The last variable should be `cpuset` and not `"cpuset"` (without quotes)

swatisehgal replied:
Thanks, will update now.

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -450,7 +452,7 @@ func (m *manager) reconcileState() (success []reconciledContainer, failure []rec
 				_, _, err := m.containerMap.GetContainerRef(containerID)
 				if err == nil {
-					klog.V(4).InfoS("ReconcileState: ignoring terminated container", "pod", klog.KObj(pod), "containerID", containerID)
+					klog.V(4).InfoS("ReconcileState: terminated container ignored", "pod", klog.KObj(pod), "containerID", containerID)
```
nit: `ignoring terminated container` ?
your call :)

swatisehgal replied:
Was attempting to follow the recommendations in the log message styling guidelines [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#message-style-guidelines) and change to past tense :) If you prefer to leave it the way it is, I don't mind :)

ffromani replied:
If you didn't change already, leave as it is. It's really a nit

ffromani reviewed (commented):
The only open point IMO is the `UID` vs `namespace/name` approach. Let's see what we have now:
```
~/go/src/k8s.io/kubernetes/pkg/kubelet/cm$ git grep klog | grep KObj | wc -l
53
~/go/src/k8s.io/kubernetes/pkg/kubelet/cm$ git grep klog | grep podUID | wc -l
38
~/go/src/k8s.io/kubernetes/pkg/kubelet/cm$ git grep klog | grep KObj | grep podUID | wc -l
4
```

So we prefer to use `KObj` aka `namespace/name` over `uid`, but:
1. there are places on which we only have `podUID` because of how the API looks like
2. there are likely few places on which we have both handy

Considering we have pending work about contextual logging and considering that it's always easier to add info rather than remove, I recommend to only add `klog.KObj` in the logs, and never log `podUID` unless it's the only identifying item we have

I would defer this conversation to the contextual logging change, where we can carry both and we can also play with verbosity level, because this approach was already used in the project in the past: https://kubernetes.slack.com/archives/C20HH14P7/p1737563609490089

swatisehgal commented:
> Considering we have pending work about contextual logging and considering that it's always easier to add info rather than remove, I recommend to only add `klog.KObj` in the logs, and never log `podUID` unless it's the only identifying item we have

Sure, I am happy with this approach! I will update this PR (and other ones) taking this into consideration.

ffromani commented:
@bart0sh any further comments? LGTM from my side

swatisehgal commented:
/test pull-kubernetes-e2e-gce
unrelated failure

bart0sh commented:
/lgtm

/assign @SergeyKanzhelev @mrunalp 
for approval

ffromani commented:
/approve
